modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm

```diff
@@ -0,0 +1,412 @@
+
+	switch(diff)
+		if(MINESWEEPER_BEGINNER) // 10x10, 10 mines
+			width = 10
+			height = 10
+			mines = 10
+		if(MINESWEEPER_INTERMEDIATE) // 17x17, 40 mines
+			width = 17
+			height = 17
+			mines = 40
+		if(MINESWEEPER_EXPERT) // 30x16, 99 mines
+			width = 30
+			height = 16
+			mines = 99
+		if(MINESWEEPER_CUSTOM)
+			width = custom_width
+			height = custom_height
+			mines = custom_mines
+
+	tiles_left = width * height
+
+	mines = min(FLOOR(tiles_left/2,1), mines) // Crash protection
+
+	for(var/i=1, i<mines+1, i++) // Set up mines
+		var/mine_spot = list(rand(0,width-1),rand(0,height-1))
+		while(find_in_mines(mine_spot)) // There's already a mine here! Choose another spot
+			mine_spot = list(rand(0,width-1),rand(0,height-1))
+		mine_spots += list(mine_spot)
+
+	for(var/y=1, y<height+1, y++) // Set all squares to be hidden
+		for(var/x=1, x<width+1, x++)
+			board_data[x][y] = "minesweeper_hidden.png"
+
+	game_status = MINESWEEPER_CONTINUE
+
+/datum/minesweeper/proc/select_square(x,y)
+	if(find_in_mines(list(x-1,y-1)))
+		board_data[x][y] = "minesweeper_minehit.png"
+		for(var/list/mine in mine_spots)
+			if(mine[1] == x-1 && mine[2] == y-1)
+				continue
+			board_data[mine[1]+1][mine[2]+1] = "minesweeper_mine.png"
+		switch(rand(1,3))
+			if(1)
+				play_snd('modular_zubbers/sound/arcade/minesweeper_explosion1.ogg')
+			if(2)
+				play_snd('modular_zubbers/sound/arcade/minesweeper_explosion2.ogg')
+			if(3)
+				play_snd('modular_zubbers/sound/arcade/minesweeper_explosion3.ogg')
+		return MINESWEEPER_DEAD
+
+	tiles_left--
+
+	var/mine_count = 0
+	for(var/scanx=-1, scanx<2, scanx++) // -1, 0, 1
+		for(var/scany=-1, scany<2, scany++)
+			if(scanx+x < 1 || scany+y < 1)
+				continue
+			if(scanx == 0 && scany == 0) // We know we aren't a mine
+				continue
+			if(board_data[scanx+x][scany+y] != "minesweeper_hidden.png" && board_data[scanx+x][scany+y] != "minesweeper_flag.png")
+				continue
+			if(find_in_mines(list(scanx+x-1,scany+y-1)))
+				mine_count++
+
+	if(mine_count == 0) // There are no mines around me! Select every square adjacent!
+		board_data[x][y] = "minesweeper_empty.png"
+		for(var/scanx=-1, scanx<2, scanx++) // -1, 0, 1
+			for(var/scany=-1, scany<2, scany++)
+				if(scanx+x < 1 || scany+y < 1)
+					continue
+				if(scanx == 0 && scany == 0)
+					continue
+				if(board_data[scanx+x][scany+y] != "minesweeper_hidden.png")
+					continue
+				select_square(scanx+x,scany+y)
+	else
+		board_data[x][y] = "minesweeper_[mine_count].png"
+
+	return tiles_left <= mines ? MINESWEEPER_VICTORY : MINESWEEPER_CONTINUE
+
+/datum/minesweeper/proc/diff_text(diff)
+	return list("Beginner", "Intermediate", "Expert", "Custom")[diff]
+
+/datum/minesweeper/proc/find_in_mines(list/coord)
+	var/order = 0
+	for(var/list/L in mine_spots)
+		order++
+		if(coord[1] == L[1] && coord[2] == L[2])
+			return order
+	return FALSE
+
+/datum/minesweeper/proc/is_blank_tile_start(x,y)
+	for(var/scanx=-1, scanx<2, scanx++) // -1, 0, 1
```

Comment:
theres a microopt to be had here

Reply:
No clue what this is meant to be about here

Reply:
The fuck is a microopt

Reply:
```
for(var/i=-1 to limit)
    dothing()
```

runs faster than a traditional javascript style loop